*** gropdf.old	Wed Aug 21 16:17:01 2019
--- gropdf	Wed Aug 21 16:17:01 2019
***************
*** 22,27 ****
--- 22,31 ----
  
  use strict;
  use Getopt::Long qw(:config bundling);
+ use Encode;
+ 
+ my $use_suppl_font = 1;
+ my $use_unicode_bookmark = 1;
  
  use constant
  {
***************
*** 30,35 ****
--- 34,40 ----
      PSNAME		=> 2,
      ASSIGNED		=> 3,
      USED		=> 4,
+     SUPPL		=> 5,
  };
  
  my $gotzlib=0;
***************
*** 79,84 ****
--- 84,90 ----
  my $stream='';	# Current Text/Graphics stream
  my $cftsz=10;	# Current font sz
  my $cft;	# Current Font
+ my $cftsup=0;	# Current Font (supplemental)
  my $lwidth=1;	# current linewidth
  my $linecap=1;
  my $linejoin=1;
***************
*** 390,396 ****
  	push(@{$o->{WIDTH}},$wid);
  	last if $#{$o->{DIFF}} >= 255;
      }
!     unshift(@{$o->{DIFF}},0);
      my $p=GetObj($fontlst{$fontno}->{OBJ});
  
      if (exists($p->{LastChar}) and $p->{LastChar} > 255)
--- 396,402 ----
  	push(@{$o->{WIDTH}},$wid);
  	last if $#{$o->{DIFF}} >= 255;
      }
!     unshift(@{$o->{DIFF}},0) if !$use_suppl_font;
      my $p=GetObj($fontlst{$fontno}->{OBJ});
  
      if (exists($p->{LastChar}) and $p->{LastChar} > 255)
***************
*** 399,404 ****
--- 405,444 ----
  	splice(@{$o->{DIFF}},256);
  	splice(@{$o->{WIDTH}},256);
      }
+ 
+     if ($use_suppl_font) {
+ 	my $fnt = $o;
+ 	while ($fnt = $fnt->{NEXT}) {
+ 	    my (@d, @w);
+ 
+ 	    foreach my $cn (0..255) {
+ 		my $ch = $fnt->{NO}->[$cn + $fnt->{SUPPL} * 256];
+ 		if ($ch->[0] && $ch->[1] && $ch->[0] ne $ch->[1]) {
+ 		    my $g = $ch->[1];
+ 		    push @d, $fnt->{NAM}->{$g}->[PSNAME] || '/.notdef';
+ 		    push @w, $fnt->{NAM}->{$g}->[WIDTH] || 0;
+ 		} else {
+ 		    push @d, '/.notdef';
+ 		    push @w, 0;
+ 		}
+ 	    }
+ 
+ 	    my $obj = BuildObj($objct + 1, {
+ 		%{$p}{qw/Type Subtype BaseFont FontDescriptor/},
+ 		Widths => \@w,
+ 		FirstChar => 0,
+ 		LastChar => 255,
+ 		Encoding => BuildObj($objct + 2, {
+ 		    Type => '/Encoding',
+ 		    Differences => \@d,
+ 		}),
+ 	    });
+ 	    $objct += 2;
+ 
+ 	    my $q = GetObj(2);
+ 	    $q->{Resources}->{Font}->{$fnt->{NM}.'.'.$fnt->{SUPPL}} = $obj;
+ 	}
+     }
  }
  
  foreach my $o (3..$objct)
***************
*** 902,907 ****
--- 942,954 ----
  		    $t=~s/\\e/\\\\/g;
  		    $t=~m/(^.*\/Title \()(.*)(\).*)/;
  		    my ($pre,$title,$post)=($1,$2,$3);
+ 		    if ($use_unicode_bookmark &&
+ 			$title =~ s/\\\[u([0-9A-F_]+)\]/join(
+ 			    '', map { pack "U", hex } split '_', $1
+ 			  )/eg) {
+ 			$title = join '', map sprintf("\\%o", $_),
+ 			  unpack "C*", encode("utf16", $title);
+ 		    }
  		    $title=~s/(?<!\\)\(/\\\(/g;
  		    $title=~s/(?<!\\)\)/\\\)/g;
  		    my @xwds=split(' ',"<< $pre$title$post >>");
***************
*** 2191,2196 ****
--- 2238,2251 ----
      my $ascent=0;
      my $charset='';
  
+     $fnt{NM} = 'F'.$fontno;
+     $fnt{SUPPL} = 0;
+     my @remap = (128..138, 145..255); # ignore ligatures. see text.enc.
+     $fnt{REMAP} = \@remap;
+     my @used;
+     $used[$_] = 1 for 0..255;
+     $used[$_] = 0 for @remap;
+ 
      while (<$f>)
      {
  	chomp;
***************
*** 2230,2236 ****
--- 2285,2297 ----
  	    $r[0]='u0020' if $r[3] == 32;
  	    $r[0]="u00".hex($r[3]) if $r[0] eq '---';
  #	    next if $r[3] >255;
+ 	    if ($fnt{NAM}->{$r[0]}) {
+ 		#Msg(0, "$r[0], $r[3], /$r[4] - dup in $ofontnm") if $debug;
+ 		next;
+ 	    }
  	    $fnt{NAM}->{$r[0]}=[$p[0],$r[3],'/'.$r[4],$r[3],0];
+ 	    $fnt{NAM}->{$r[0]}->[SUPPL] = 0;
+ 	    $fnt{NAM}->{$r[0]}->[USED] = $used[$r[3]];
  	    $fnt{NO}->[$r[3]]=[$r[0],$r[0]];
  	    $lastnm=$r[0];
  	    $lastchr=$r[3] if $r[3] > $lastchr;
***************
*** 2258,2263 ****
--- 2319,2325 ----
      $fnt{DIFF}=[];
      $fnt{WIDTH}=[];
      $fnt{NAM}->{''}=[0,-1,'/.notdef',-1,0];
+     $fnt{NAM}->{''}->[SUPPL] = 0;
      $slant=-$fnt{'slant'} if exists($fnt{'slant'});
      $fnt{'spacewidth'}=700 if !exists($fnt{'spacewidth'});
  
***************
*** 2307,2314 ****
  			);
  
  	$objct+=3;
! 	$fontlst{$fontno}->{NM}='/F'.$fontno;
! 	$pages->{'Resources'}->{'Font'}->{'F'.$fontno}=$fontlst{$fontno}->{OBJ};
  	$fontlst{$fontno}->{FNT}=\%fnt;
  	$obj[$objct]->{STREAM}=$t1stream;
  
--- 2369,2376 ----
  			);
  
  	$objct+=3;
! 	$fontlst{$fontno}->{NM}='/'.$fnt{NM};
! 	$pages->{'Resources'}->{'Font'}->{$fnt{NM}}=$fontlst{$fontno}->{OBJ};
  	$fontlst{$fontno}->{FNT}=\%fnt;
  	$obj[$objct]->{STREAM}=$t1stream;
  
***************
*** 2345,2352 ****
  			);
  
  	$objct+=2;
! 	$fontlst{$fontno}->{NM}='/F'.$fontno;
! 	$pages->{'Resources'}->{'Font'}->{'F'.$fontno}=$fontlst{$fontno}->{OBJ};
  	$fontlst{$fontno}->{FNT}=\%fnt;
      }
  
--- 2407,2414 ----
  			);
  
  	$objct+=2;
! 	$fontlst{$fontno}->{NM}='/'.$fnt{NM};
! 	$pages->{'Resources'}->{'Font'}->{$fnt{NM}}=$fontlst{$fontno}->{OBJ};
  	$fontlst{$fontno}->{FNT}=\%fnt;
      }
  
***************
*** 2378,2383 ****
--- 2440,2446 ----
      Msg(1,"Failed to open '$file'") if !defined($f);
  
      $head=GetChunk($f,1,"currentfile eexec");
+     $body=$tail='';
      $body=GetChunk($f,2,"00000000") if !eof($f);
      $tail=GetChunk($f,3,"cleartomark") if !eof($f);
  
***************
*** 2590,2595 ****
--- 2653,2659 ----
  
  #	IsText();
      $cft="$par";
+     $cftsup=0;
      $fontchg=1;
  #	$stream.="/F$cft $cftsz Tf\n" if $cftsz;
      $widtbl=CacheWid($par);
***************
*** 2646,2652 ****
  	if (defined($cft))
  	{
  	    $whtsz=$fontlst{$cft}->{FNT}->{spacewidth}*$cftsz;
! 	    $stream.="/F$cft $cftsz Tf\n";
  	}
  	$stream.="$curkern Tc\n";
      }
--- 2710,2718 ----
  	if (defined($cft))
  	{
  	    $whtsz=$fontlst{$cft}->{FNT}->{spacewidth}*$cftsz;
! 	    $stream.="/F$cft";
! 	    $stream.=".$cftsup" if $cftsup;
! 	    $stream.=" $cftsz Tf\n";
  	}
  	$stream.="$curkern Tc\n";
      }
***************
*** 2663,2671 ****
      if ($fontchg)
      {
  	PutLine(0);
! 	$whtsz=$fontlst{$cft}->{FNT}->{spacewidth}*$cftsz;
! 	$stream.="/F$cft $cftsz Tf\n" if $cftsz and defined($cft);
! 	$fontchg=0;
      }
  
      $mode='t';
--- 2729,2742 ----
      if ($fontchg)
      {
  	PutLine(0);
! 	if (defined($cft))
! 	{
! 	    $whtsz=$fontlst{$cft}->{FNT}->{spacewidth}*$cftsz;
! 	    $stream.="/F$cft";
! 	    $stream.=".$cftsup" if $cftsup;
! 	    $stream.=" $cftsz Tf\n";
! 	    $fontchg=0;
! 	}
      }
  
      $mode='t';
***************
*** 3302,3307 ****
--- 3373,3379 ----
  sub TextWid
  {
      my $txt=shift;
+     my $sup=shift;
      my $fnt=shift;
      my $w=0;
      my $ck=0;
***************
*** 3309,3314 ****
--- 3381,3387 ----
      foreach my $c (split('',$txt))
      {
  	my $cn=ord($c);
+ 	$cn+=$sup*256;
  	$widtbl->[$cn]=$origwidtbl->[$cn]*$cftsz if !defined($widtbl->[$cn]);
  	$w+=$widtbl->[$cn];
      }
***************
*** 3322,3327 ****
--- 3395,3403 ----
  {
      my $par=shift;
      my $fnt=$fontlst{$cft}->{FNT};
+     my $sup = shift || 0;
+     $fontchg=1 if $cftsup != $sup;
+     $cftsup = $sup;
  
      if ($kernadjust != $curkern)
      {
***************
*** 3336,3355 ****
      foreach my $j (0..length($par2)-1)
      {
  	my $cn=ord(substr($par2,$j,1));
! 	my $chnm=$fnt->{NAM}->{$fnt->{NO}->[$cn]->[1]};
  
! 	if ($chnm->[USED]==0)
! 	{
! 	    $chnm->[USED]=1;
! 	}
! 	elsif ($fnt->{NO}->[$cn]->[0] ne $fnt->{NO}->[$cn]->[1])
  	{
! 	    # A glyph has already been remapped to this char, so find a spare
! 
! 	    my $cn2=RemapChr($cn,$fnt,$fnt->{NO}->[$cn]->[0]);
! 	    $stream.="% MMM Remap $cn to $cn2\n" if $debug;
  
! 	    if ($cn2)
  	    {
  		substr($par2,$j,1)=chr($cn2);
  
--- 3412,3427 ----
      foreach my $j (0..length($par2)-1)
      {
  	my $cn=ord(substr($par2,$j,1));
! 	$cn+=$sup*256;
! 	my $chnm=$fnt->{NO}->[$cn]->[1];
  
! 	if (!$fnt->{NAM}->{$chnm}->[USED])
  	{
! 	    my ($cn2, $sup2) = RemapChr($cn, $fnt, $chnm);
! 	    $stream.="% MMM Remap $cn,$sup to $cn2,$sup2\n" if $debug;
! 	    Msg(0, "got: $fnt->{NM}.$sup2; expected: $fnt->{NM}.$sup\n") if $sup != $sup2;
  
! 	    #if ($cn2)
  	    {
  		substr($par2,$j,1)=chr($cn2);
  
***************
*** 3364,3370 ****
  	    }
  	}
      }
!     my $wid=TextWid($par2,$fnt);
  
      $par=reverse(split('',$par)) if $xrev and $par!~m/^!\|!\|(\d\d\d)/;
  
--- 3436,3442 ----
  	    }
  	}
      }
!     my $wid=TextWid($par2,$sup,$fnt);
  
      $par=reverse(split('',$par)) if $xrev and $par!~m/^!\|!\|(\d\d\d)/;
  
***************
*** 3389,3396 ****
      if ($fontchg)
      {
  	PutLine();
! 	$whtsz=$fontlst{$cft}->{FNT}->{spacewidth}*$cftsz;
! 	$stream.="/F$cft $cftsz Tf\n", $fontchg=0 if $fontchg && defined($cft);
      }
  
      $gotT=1;
--- 3461,3474 ----
      if ($fontchg)
      {
  	PutLine();
! 	if (defined($cft))
! 	{
! 	    $whtsz=$fontlst{$cft}->{FNT}->{spacewidth} * $cftsz;
! 	    $stream.="/F$cft";
! 	    $stream.=".$cftsup" if $cftsup;
! 	    $stream.=" $cftsz Tf\n";
! 	    $fontchg=0;
! 	}
      }
  
      $gotT=1;
***************
*** 3491,3499 ****
  {
      my $par=shift;
  
!     my ($par2,$nm)=FindChar($par);
! 
!     do_t($par2);
      $nomove=$fontlst{$cft}->{FNT}->{NAM}->{$par}->[WIDTH]*$cftsz ;
  }
  
--- 3569,3575 ----
  {
      my $par=shift;
  
!     do_t(FindChar($par));
      $nomove=$fontlst{$cft}->{FNT}->{NAM}->{$par}->[WIDTH]*$cftsz ;
  }
  
***************
*** 3504,3514 ****
  
      if (exists($fnt->{NAM}->{$chnm}))
      {
! 	my $ch=$fnt->{NAM}->{$chnm}->[ASSIGNED];
! 	$ch=RemapChr($ch,$fnt,$chnm) if ($ch > 255);
! 	$fnt->{NAM}->{$chnm}->[USED]=0 if $fnt->{NO}->[$ch]->[1] eq $chnm;
! 
! 	return(($ch<32)?sprintf("!|!|%03o",$ch):chr($ch),$widtbl->[$ch]);
      }
      else
      {
--- 3580,3588 ----
  
      if (exists($fnt->{NAM}->{$chnm}))
      {
! 	my ($ch,$sup,$used)=@{$fnt->{NAM}->{$chnm}}[ASSIGNED,SUPPL,USED];
! 	($ch,$sup) = RemapChr($ch,$fnt,$chnm) if !$used;
! 	return ($ch<32)? sprintf("!|!|%03o",$ch) : chr($ch), $sup;
      }
      else
      {
***************
*** 3521,3527 ****
      my $ch=shift;
      my $fnt=shift;
      my $chnm=shift;
!     my $unused=0;
  
      foreach my $un (0..$#{$fnt->{NO}})
      {
--- 3595,3634 ----
      my $ch=shift;
      my $fnt=shift;
      my $chnm=shift;
!     my $unused;
! 
!     if ($use_suppl_font) {
! 
!         while (defined(my $un = shift @{$fnt->{REMAP}})) {
!             my $ux = $un + $fnt->{SUPPL} * 256;
! 	    my $glyph = $fnt->{NO}->[$ux]->[1];
!             $unused = $un, last if !$glyph || !$fnt->{NAM}->{$glyph}->[USED];
!         }
! 
!         if (!defined $unused) {
!             if (!$fnt->{NEXT}) {
!                 my $fnt2 = {
! 		    %{$fnt}{qw/NM NO NAM/},
!                     SUPPL => $fnt->{SUPPL} + 1,
!                     REMAP => [ 0..255 ],
!                 };
!                 $fnt->{NEXT} = $fnt2;
!             }
!             return RemapChr($ch, $fnt->{NEXT}, $chnm);
!         }
! 
! 	my $ux = $unused + $fnt->{SUPPL} * 256;
! 	my $glyph = $fnt->{NO}->[$ux]->[1];
! 	delete($fontlst{$cft}->{CACHE}->{$cftsz});
! 	@{$fnt->{NAM}->{$chnm}}[ASSIGNED, SUPPL, USED] = ($unused, $fnt->{SUPPL}, 1);
! 	$fnt->{NO}->[$ux]->[1] = $chnm;
! 	$widtbl = CacheWid($cft);
! 
!         $stream .= "% AAA Assign $chnm ($ch) to $unused ($fnt->{SUPPL})\n" if $debug;
! 
! 	$ch = $unused;
! 	return ($ch, $fnt->{SUPPL});
!     }
  
      foreach my $un (0..$#{$fnt->{NO}})
      {
***************
*** 3529,3535 ****
  	$unused=$un,last if $fnt->{NO}->[$un]->[1] eq '';
      }
  
!     if (!$unused)
      {
  	foreach my $un (128..255)
  	{
--- 3636,3642 ----
  	$unused=$un,last if $fnt->{NO}->[$un]->[1] eq '';
      }
  
!     if (!defined $unused)
      {
  	foreach my $un (128..255)
  	{
***************
*** 3539,3545 ****
  	}
      }
  
!     if ($unused && $unused <= 255)
      {
  	my $glyph=$fnt->{NO}->[$unused]->[1];
  	delete($fontlst{$cft}->{CACHE}->{$cftsz});
--- 3646,3652 ----
  	}
      }
  
!     if (defined $unused && $unused <= 255)
      {
  	my $glyph=$fnt->{NO}->[$unused]->[1];
  	delete($fontlst{$cft}->{CACHE}->{$cftsz});
***************
*** 3550,3561 ****
  	$stream.="% AAA Assign $chnm ($ch) to $unused\n" if $debug;
  
  	$ch=$unused;
! 	return($ch);
      }
      else
      {
  	Msg(0,"Too many glyphs used in font '$cft'");
! 	return(32);
      }
  }
  
--- 3657,3668 ----
  	$stream.="% AAA Assign $chnm ($ch) to $unused\n" if $debug;
  
  	$ch=$unused;
! 	return($ch,0);
      }
      else
      {
  	Msg(0,"Too many glyphs used in font '$cft'");
! 	return(32,0);
      }
  }
  
