--- man.local.bak	2018-12-23 23:37:31.167309000 +0900
+++ man.local	2018-12-23 23:51:31.075920000 +0900
@@ -9999,0 +10000,10 @@
+.
+.\" reconfigure line gap.
+.if t \{\
+.  if "\*[locale]"japanese" \{\
+.am1 TH
+.  nr VS (\\n[PS] * 180 / 100)
+.  vs \\n[VS]u
+..
+.  \}
+.\}
--- mdoc.local.bak	2018-12-23 23:37:45.093011000 +0900
+++ mdoc.local	2018-12-23 23:51:40.518703000 +0900
@@ -9999,0 +10000,7 @@
+.
+.\" reconfigure line gap.
+.if t \{\
+.  if "\*[locale]"japanese" \{\
+.    vs (\n[.s]p * 180 / 100)
+.  \}
+.\}
