*** pdf.tmac.orig	2021-09-19 22:47:31.433533814 +0000
--- pdf.tmac	2021-09-19 22:48:36.232725972 +0000
***************
*** 250,256 ****
  .         chop pdf:clean
  .         asciify pdf:clean
  .         length pdf:clean:len \\*[pdf:clean]
! .         ds pdf:cleaned \\*[pdf:clean]
  .         rm pdf:clean
  .         ev
  .         tr \[em]\[em]
--- 250,260 ----
  .         chop pdf:clean
  .         asciify pdf:clean
  .         length pdf:clean:len \\*[pdf:clean]
! .   \" 
! .   \" use pdf:cleaned instead of pdf:clean for now
! .   \" because unicode characters ignored in bookmarks. xxxxx
! .   \"
! .         \"ds pdf:cleaned \\*[pdf:clean]
  .         rm pdf:clean
  .         ev
  .         tr \[em]\[em]
